src/Oscoin/Protocol/Sync.hs

```diff
@@ -0,0 +1,393 @@
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Oscoin.Protocol.Sync
+    (
+    -- * Types
+      Sync(..)
+    , SyncError(..)
+    , Timeout(..)
+    , ActivePeer
+    , ActivePeers
+    , SyncEvent(..)
+    , SyncContext(..)
+    , DataFetcher(..)
+    , ProtocolRequest(..)
+    , SProtocolRequest(..)
+    , ProtocolResponse
+    , Range(..)
+
+    -- * Pure functions
+    , range
+    , isDone
+
+    -- * Syncing functions
+    , syncBlocks
+
+    -- * Testing internals
+    , withActivePeers
+    , withActivePeer
+    , getRemoteTip
+    , getRemoteHeight
+    ) where
+
+import           Oscoin.Crypto.Hash (Hash)
+import           Oscoin.Crypto.PubKey (PublicKey)
+import           Oscoin.Prelude
+import           Prelude (last)
+
+import           Oscoin.Consensus.Nakamoto (PoW)
+
+import           Oscoin.Crypto.Blockchain.Block
+                 (Block, BlockHeader, Height, Sealed, blockHeader, blockHeight)
+import           Oscoin.Data.Tx
+import           Oscoin.P2P as P2P
+import           Oscoin.Storage.Block.Abstract (BlockStoreReader)
+import           Oscoin.Telemetry.Trace
+import           Oscoin.Time (Duration)
+import           Oscoin.Time.Chrono
+
+import           Data.HashSet (HashSet)
+import qualified Data.HashSet as HS
+import           Data.List ((\\))
+import           Data.List.Split (chunksOf)
+import qualified Data.Map.Strict as M
+import qualified Data.Set as Set
+import           GHC.Natural
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+type ActivePeer c  = NodeInfo c
+type ActivePeers c = HashSet (ActivePeer c)
+
+data SyncEvent c tx s =
+      SyncBlock (Block c tx (Sealed c s))
+    | SyncBlockHeader (BlockHeader c s)
+
+deriving instance ( Eq (Hash c)
+                  , Eq tx
+                  , Eq s
+                  , Eq (PublicKey c)
+                  ) => Eq (SyncEvent c tx s)
+deriving instance ( Ord (Hash c)
+                  , Ord tx
+                  , Ord s
+                  , Ord (PublicKey c)
+                  ) => Ord (SyncEvent c tx s)
+deriving instance ( Show (Hash c)
+                  , Show tx
+                  , Show s
+                  , Show (PublicKey c)
+                  ) => Show (SyncEvent c tx s)
+
+data SyncError =
+      RequestTimeout ProtocolRequest Addr Timeout
+      -- ^ The given peer exceeded the request timeout when serving this
+      -- 'SyncProtocolRequest'.
+    | NoActivePeers
+      -- ^ There are not active peers to talk to.
+    | NoRemoteTipFound
+      -- ^ It was not possible to fetch a valid tip from our peers.
+    | AllPeersTimeoutError
+      -- ^ All the queried peers didn't respond in the allocated timeout.
+    deriving (Show, Eq)
+
+instance Exception SyncError
+
+data Timeout =
```

Comment:
Are timeouts ever considered? If not, then why have a "checked" error at all?

Reply:
I originally was thinking to have `timed` being a standalone combinator, but looks like we might get away by simply merging `Timeout` into `RequestTimeout`.